src/utilities/substituteFragmentArguments.ts

```diff
@@ -0,0 +1,78 @@
+import type { Maybe } from '../jsutils/Maybe.js';
+import type { ObjMap } from '../jsutils/ObjMap.js';
+
+import type {
+  ArgumentNode,
+  FragmentDefinitionNode,
+  FragmentSpreadNode,
+  SelectionSetNode,
+  ValueNode,
+  VariableDefinitionNode,
+} from '../language/ast.js';
+import { Kind } from '../language/kinds.js';
+import { visit } from '../language/visitor.js';
+
+/**
+ * Replaces all fragment argument values with non-fragment-scoped values.
+ *
+ * NOTE: fragment arguments are scoped to the fragment they're defined on.
+ * Therefore, after we apply the passed-in arguments, all remaining variables
+ * must be either operation defined variables or explicitly unset.
+ */
+export function substituteFragmentArguments(
+  def: FragmentDefinitionNode,
+  fragmentSpread: FragmentSpreadNode,
+): SelectionSetNode {
+  const argumentDefinitions = def.variableDefinitions;
+  if (argumentDefinitions == null || argumentDefinitions.length === 0) {
+    return def.selectionSet;
+  }
+  const argumentValues = fragmentArgumentSubstitutions(
+    argumentDefinitions,
+    fragmentSpread.arguments,
+  );
+  return visit(def.selectionSet, {
+    Variable(node) {
+      return argumentValues[node.name.value];
+    },
+  });
+}
+
+export function fragmentArgumentSubstitutions(
+  variableDefinitions: ReadonlyArray<VariableDefinitionNode>,
+  argumentValues: Maybe<ReadonlyArray<ArgumentNode>>,
+): ObjMap<ValueNode> {
+  const substitutions: ObjMap<ValueNode> = {};
+  if (argumentValues) {
+    for (const argument of argumentValues) {
+      substitutions[argument.name.value] = argument.value;
+    }
+  }
+
+  for (const variableDefinition of variableDefinitions) {
+    const argumentName = variableDefinition.variable.name.value;
+    if (substitutions[argumentName]) {
+      continue;
+    }
+
+    const defaultValue = variableDefinition.defaultValue;
+    if (defaultValue) {
+      substitutions[argumentName] = defaultValue;
+    } else {
+      // We need a way to allow unset arguments without accidentally
+      // replacing an unset fragment argument with an operation
+      // variable value. Fragment arguments must always have LOCAL scope.
+      //
+      // To remove this hack, we need to either:
+      //    - include fragment argument scope when evaluating fields
+      //    - make unset fragment arguments invalid
+      // Requiring the spread to pass all non-default-defined arguments is nice,
+      // but makes field argument default values impossible to use.
+      substitutions[argumentName] = {
+        kind: Kind.VARIABLE,
+        name: { kind: Kind.NAME, value: '__UNSET' },
+      };
```

Comment:
> ```
> // To remove this hack, we need to either:
> //   - include fragment argument scope when evaluating fields
> ```

I think probably yes, that. Is that too crazy? This hack is not ideal.

Also, `query ($__UNSET: String = "SomeDefault") {...` is valid and while most query authors wouldn't write this, this seems like some kind of security vector or something

Reply:
Oh I should just replace it with `$0__UNSET` and then it's not valid :)

> Is that too crazy?

I spent a week or so of full-time work trying to get the spec algorithms and graphql-js to have full scope. It was just a much more pervasive change. It's not too crazy, but the replacement strategy is certainly a *smaller* change for both.

Reply:
fair enough. That makes me think we should call this "Fragment Templates" and mirror some of the C++ nomenclature. I don't think I like it, but it's closer to what we're describing :)

would an equivalent behavior be to also visit each InputValue on leave and if its unset to remove it? In other words - rather than putting in a hacked variable AST node, leave the AST that would be the equivalent to an unset var